### DISCUSSION TITLE
Burglar Alarm

### DISCUSSION THREAD

**Discussion by adamlilley88** [HomeSpan/HomeSpan#790] [Q&A]
I am very new to coding.

I am trying to build a replacement to my wired alarm system that incorporates into HomeKit

I have in total:

Digital Inputs: (I would like these to show as contact and motion sensors in HomeKit so that I can use custom automations also)
2x Door Contact Sensors
5x PIR Motion Sensors

Digital Outputs
1x Siren

I have tried various examples and I am struggling to compile something that works, any help would be appreciated

**Comment by frankonski** [comment]
What is your alarm control panel make and model?

**Reply by adamlilley88** [reply]
Its an ESP32 device, I am replacing by 'dumb' alarm device completely.

I have successfully managed to incorporate the existing PIR and contact sensors into Homespan.

I would like to add in a 'Security System' that can be set through HomeKit, to poll certain sensors (depending on Away/Night setting) for motion and trigger the alarm

And when the alarm is triggered, set an output to drive the physical siren on the house

**Comment by frankonski** [comment]
Check the example at https://github.com/taligentx/dscKeybusInterface/blob/master/examples/esp32/HomeKit-HomeSpan/HomeKit-HomeSpan.ino, that gives an example of using SecuritySystem.

**Reply by adamlilley88** [reply]
yes but this is to integrate with an existing burglar alarm system.

I am wanting to configure the whole system on my ESP32

**Reply by frankonski** [reply]
Correct, but all you need to do is implement your own structure based on `Service::SecuritySystem`, and implement your logic in `loop()` to get the status of all of your sensors and change characteristic values, and implement `update()` to handle actions coming from an app.

**Reply by adamlilley88** [reply]
I think this is where it is breaking down, I can't seem to get a successful compile.

If I send you the code somehow, can you take a look?

**Reply by adamlilley88** [reply]
```
#include "HomeSpan.h"
#include "Contact.h"
#include "Motion.h"

struct SecuritySystem : Service::SecuritySystem {


  Characteristic::SecuritySystemCurrentState current{0};
  Characteristic::SecuritySystemTargetState target{0};

  SecuritySystem()
    : Service::SecuritySystem() {
    target.setValidValues(3, 1, 2);
  }
SecuritySystem *securitySystem;

  boolean update() override {
    current.setVal(target.getNewVal());
    return (true);
}
};

void setup() {

  Serial.begin(115200);
int Siren = 14;
bool WarningDelay = 0;
int WarningSound = 27;
  pinMode(Siren, OUTPUT);
  pinMode(WarningSound, OUTPUT);


  homeSpan.begin(Category::Bridges, "Home Alarm Bridge");

  new SpanAccessory();
  new Service::AccessoryInformation();
  new Characteristic::Identify();
  new Characteristic::Manufacturer("Test");
  new Characteristic::SerialNumber("2402001");
  new Characteristic::Model("1.0");
  new Characteristic::FirmwareRevision("1.0");

  new SpanAccessory();
  new Service::AccessoryInformation();
  new Characteristic::Identify();
  new Characteristic::Name("Front Door");
  new DEV_ContactSensor(19);

  new SpanAccessory();
  new Service::AccessoryInformation();
  new Characteristic::Identify();
  new Characteristic::Name("Side Door");
  new DEV_ContactSensor(18);

  new SpanAccessory();
  new Service::AccessoryInformation();
  new Characteristic::Identify();
  new Characteristic::Name("Living Room PIR");
  new DEV_MotionSensor(5);

  new SpanAccessory();
  new Service::AccessoryInformation();
  new Characteristic::Identify();
  new Characteristic::Name("Play Room PIR");
  new DEV_MotionSensor(17);

  new SpanAccessory();
  new Service::AccessoryInformation();
  new Characteristic::Identify();
  new Characteristic::Name("Hallway PIR");
  new DEV_MotionSensor(16);

  new SpanAccessory();
  new Service::AccessoryInformation();
  new Characteristic::Identify();
  new Characteristic::Name("Kitchen PIR");
  new DEV_MotionSensor(4);

  new SpanAccessory();
  new Service::AccessoryInformation();
  new Characteristic::Identify();
  new Characteristic::Name("Landing PIR");
  new DEV_MotionSensor(0);


}

void loop() {
  homeSpan.poll();

  if (securitySystem.getValue(1 || 2)) && digitalRead(19 == LOW)) {
      digitalWrite(WarningDelay, 1);
    }

  if (securitySystem->current.getNewVal(1) || securitySystem->current.getNewVal(2) && digitalRead(18 == LOW)) {
      digitalWrite(WarningDelay, 1);
    }

  if (securitySystem->current.getNewVal(1) || securitySystem->current.getNewVal(2) && digitalRead(5 == LOW)) {
      digitalWrite(WarningDelay, 1);
    }

  if (securitySystem->current.getNewVal(1) || securitySystem->current.getNewVal(2) && digitalRead(17 == LOW)) {
      digitalWrite(WarningDelay, 1);
    }

  if (securitySystem->current.getNewVal(1) || securitySystem->current.getNewVal(2) && digitalRead(16 == LOW)) {
      digitalWrite(WarningDelay, 1);
    }

  if (securitySystem->current.getNewVal(1) || securitySystem->current.getNewVal(2) && digitalRead(4 == LOW)) {
      digitalWrite(WarningDelay, 1);
    }

  if (securitySystem->current.getNewVal(2) && digitalRead(0 == LOW)) {
      digitalWrite(WarningDelay, 1);   
    }

if (digitalRead(WarningDelay == 1)) {
    digitalWrite(WarningSound, HIGH);
    delay(20000);
    digitalWrite(WarningDelay, 0);
    digitalWrite(WarningSound, LOW);
    securitySystem->current.setVal(4);
}

}
```

**Comment by frankonski** [comment]
Not sure what your intent is with `getValue(1 || 2)`… and there should not be any sleep (`delay()`) either.
All the contact and motion sensor logic should be in their respective class files and implement the `loop()` method, while will automatically get called by homespan. This is where you should do these `digitalRead()`.

Good Luck!

**Reply by adamlilley88** [reply]
So in my IF statements, I am trying to check the status of the alarm either AWAY or NIGHT mode, and if the sensors detect, then to trigger the alarm, but with a time delay to disable unwanted triggers (returning home or accidentally forgetting to disarm in the night), but I am not sure how to get the correct statements

Here is my file for the contact sensors - as I mentioned, I am very new to this so I am not sure what I need to change

```
struct DEV_ContactSensor : Service::ContactSensor {

  int SwitchPin;
  SpanCharacteristic *contactSensorState;

  DEV_ContactSensor(int SwitchPin)
    : Service::ContactSensor() {
    contactSensorState = new Characteristic::ContactSensorState();
    this->SwitchPin = SwitchPin;
    pinMode(SwitchPin, INPUT_PULLUP);
  }


  void loop() {

    if (digitalRead(SwitchPin) == HIGH) {
      contactSensorState->setVal(true);
    } else {
      contactSensorState->setVal(false);
      delay(100);
    }
  }
};

```

**Reply by frankonski** [reply]
I think you need to look back at your [operator precedence](https://en.cppreference.com/w/cpp/language/operator_precedence) as you have a mix of `&&` and `||` which might not be properly grouped based on what you are trying to achieve.

You are also passing parameters to methods that don't accept any. See the documentation of `getVal<T>()` in the [API Reference Guide](https://github.com/HomeSpan/HomeSpan/blob/master/docs/Reference.md).

**Reply by adamlilley88** [reply]
Below, I am trying to use the security system status as the condition of my IF statements. If the alarm status is either 'away' OR 'night' AND the sensor connected to pin 19 is LOW, set the alarm status to 'triggered'

I want the alarm system to have redundancy, so that in the event that HomeKit fails for whatever reason, the alarm system will still function.

```
  if (securitySystem->current.getVal<>() == 1 || securitySystem->getVal<>() == 2 ) && digitalRead(19 == LOW)) {
      digitalWrite(WarningDelay, 1);
    }
```

**Comment by adamlilley88** [comment]
So,, I finally managed to get this to compile in the Arduino IDE

However now, I am getting this boot up error


> Guru Meditation Error: Core  1 panic'ed (LoadProhibited). Exception was unhandled.
> 
> Core  1 register dump:
> PC      : 0x400d3c7b  PS      : 0x00060830  A0      : 0x800d3d83  A1      : 0x3ffceac0  
> A2      : 0x00000050  A3      : 0x00000060  A4      : 0x3ffd09d8  A5      : 0x3ffd0abc  
> A6      : 0x00000001  A7      : 0x3ffbc964  A8      : 0x800dd62e  A9      : 0x3ffce9e0  
> A10     : 0x911304ab  A11     : 0x911304ab  A12     : 0x3ffc5540  A13     : 0x3ffc48f4  
> A14     : 0x3ffbbb14  A15     : 0x3ffc45a8  SAR     : 0x0000000a  EXCCAUSE: 0x0000001c  
> EXCVADDR: 0x0000006c  LBEG    : 0x4008476d  LEND    : 0x40084775  LCOUNT  : 0x00000027


Code

```
#include "HomeSpan.h"
#include "Contact.h"
#include "Motion.h"

struct SecuritySystem : Service::SecuritySystem {


  Characteristic::SecuritySystemCurrentState current{3};
  Characteristic::SecuritySystemTargetState target{3};

  SecuritySystem()
    : Service::SecuritySystem() {
    target.setValidValues(3, 3, 1, 2);
  }


  boolean update() override {
    current.setVal(target.getNewVal());
    return (true);
}
};

SecuritySystem *securitySystem;

int FrontDoor = 19;
int SideDoor = 18;
int LivingRoomPIR = 5;
int PlayRoomPIR = 17;
int HallwayPIR = 16;
int KitchenPIR = 4;
int LandingPIR = 0;
int Siren = 14;
bool WarningActive = false;
int WarningSound = 12; 
int NightReqPin = 21;
int AwayReqPin = 22;
int TamperPin = 23;
unsigned int WarningDelay;

void setup() {

  Serial.begin(115200);

  pinMode(Siren, OUTPUT);
  pinMode(WarningSound, OUTPUT);
  pinMode(NightReqPin, INPUT_PULLUP);
  pinMode(AwayReqPin, INPUT_PULLUP);
  pinMode(TamperPin, INPUT_PULLUP);

  homeSpan.begin(Category::Bridges, "Home Alarm Bridge");

  new SpanAccessory();
  new Service::AccessoryInformation();
  new Characteristic::Identify();
  new Characteristic::Manufacturer("Test");
  new Characteristic::SerialNumber("2402001");
  new Characteristic::Model("1.0");
  new Characteristic::FirmwareRevision("1.0");

  new SpanAccessory();
  new Service::AccessoryInformation();
  new Characteristic::Identify();
  new Characteristic::Name("Front Door");
  new DEV_ContactSensor(FrontDoor);

  new SpanAccessory();
  new Service::AccessoryInformation();
  new Characteristic::Identify();
  new Characteristic::Name("Side Door");
  new DEV_ContactSensor(SideDoor);

  new SpanAccessory();
  new Service::AccessoryInformation();
  new Characteristic::Identify();
  new Characteristic::Name("Living Room PIR");
  new DEV_MotionSensor(LivingRoomPIR);

  new SpanAccessory();
  new Service::AccessoryInformation();
  new Characteristic::Identify();
  new Characteristic::Name("Play Room PIR");
  new DEV_MotionSensor(PlayRoomPIR);

  new SpanAccessory();
  new Service::AccessoryInformation();
  new Characteristic::Identify();
  new Characteristic::Name("Hallway PIR");
  new DEV_MotionSensor(HallwayPIR);

  new SpanAccessory();
  new Service::AccessoryInformation();
  new Characteristic::Identify();
  new Characteristic::Name("Kitchen PIR");
  new DEV_MotionSensor(KitchenPIR);

  new SpanAccessory();
  new Service::AccessoryInformation();
  new Characteristic::Identify();
  new Characteristic::Name("Landing PIR");
  new DEV_MotionSensor(LandingPIR);


}

void loop() {
  homeSpan.poll();

  if ((securitySystem->current.getVal() == 1 || securitySystem->current.getVal()) == 2 && digitalRead(FrontDoor == LOW)) {
      digitalWrite(WarningActive, true);
      digitalWrite(WarningSound, HIGH);
      Serial.println("Front Door, Triggered");
    }
  if ((securitySystem->current.getVal() == 1 || securitySystem->current.getVal()) == 2 && digitalRead(SideDoor == LOW)) {
      digitalWrite(WarningActive, true);
      digitalWrite(WarningSound, HIGH);
      Serial.println("Side Door, Triggered");
    }
  if ((securitySystem->current.getVal() == 1 || securitySystem->current.getVal()) == 2 && digitalRead(LivingRoomPIR == LOW)) {
      digitalWrite(WarningActive, true);
      digitalWrite(WarningSound, HIGH);
      Serial.println("Living Room Motion, Triggered");
    }
  if ((securitySystem->current.getVal() == 1 || securitySystem->current.getVal()) == 2 && digitalRead(PlayRoomPIR == LOW)) {
      digitalWrite(WarningActive, true);
      digitalWrite(WarningSound, HIGH);
      Serial.println("Play Room Motion, Triggered");
    }
  if ((securitySystem->current.getVal() == 1 || securitySystem->current.getVal()) == 2 && digitalRead(HallwayPIR == LOW)) {
      digitalWrite(WarningActive, true);
      digitalWrite(WarningSound, HIGH);
      Serial.println("Hallway Motion, Triggered");
    }
  if ((securitySystem->current.getVal() == 1 || securitySystem->current.getVal()) == 2 && digitalRead(KitchenPIR == LOW)) {
      digitalWrite(WarningActive, true);
      digitalWrite(WarningSound, HIGH);
      Serial.println("Kitchen Motion, Triggered");
    }
  if (securitySystem->current.getVal() == 1 && digitalRead(LandingPIR == LOW)) {
      digitalWrite(WarningActive, true);
      digitalWrite(WarningSound, HIGH);
      Serial.println("Landing Motion, Triggered");
    }
WarningDelay = millis();
if (digitalRead(WarningActive == true && millis() - WarningDelay > 20000))  {
    
    digitalWrite(WarningDelay, 0);
    digitalWrite(WarningSound, LOW);
    securitySystem->target.setVal(4);
    digitalWrite(WarningActive, false);

}
if (3 == securitySystem->current.getVal()) {
  Serial.println("Alarm Disarmed");
}
if (1 == securitySystem->current.getVal()) {
  Serial.println("Alarm Night Mode");
}
if (2 == securitySystem->current.getVal()) {
  Serial.println("Alarm Away Mode");
}

if (digitalRead(NightReqPin == LOW)) {
    securitySystem->target.setVal(1);
    Serial.println("Night Mode Requested");
}
if (digitalRead(AwayReqPin == LOW)) {
    securitySystem->target.setVal(2);
    Serial.println("Away Mode Requested");
}
if (digitalRead(TamperPin == LOW)) {
    securitySystem->target.setVal(4);
    Serial.println("Tamper Mode Activated");
}


};
```

**Reply by frankonski** [reply]
Best advice I can give is to use PlatformIO and use `monitor_filters = esp32_exception_decoder` in `platformio.ini`, which will automatically print the whole stack trace with line number when the code crashes. That will help you narrow down the cause of the crash in your code.

**Reply by adamlilley88** [reply]
Apparently, this is the problem

In my void loop() I have

`AlarmStatus = securitySystem->current.getVal();`

What am I doing wrong here, shouldn't that return an integer for the status of the the security system?

**Comment by frankonski** [comment]
Looking at the code above, while you have declared a pointer `SecuritySystem *securitySystem;`, you've never actually instantiated it. Which means you are trying to access null pointer, and crashed because of that.

**Reply by adamlilley88** [reply]
How do I instantiate it?
I copied that code from the HomeSpan example, but assumed it was a pointer to the structure above it?

**Reply by frankonski** [reply]
`SecuritySystem *securitySystem;` is declaring a pointer. Value will be _assumed_ as `NULL`, though good practice is to also set it to `NULL`.
`SecuritySystem *securitySystem = new SecuritySystem();` is instantiating a pointer, which should be assigned a value, if all goes well with the constructor.

Which sample did you use?

**Comment by frankonski** [comment]
Also, double-check your code.... The following code
```
if ((securitySystem->current.getVal() == 1 || securitySystem->current.getVal()) == 2 && digitalRead(FrontDoor == LOW)) {
```
will actually be interpreted as
```
if ( (securitySystem->current.getVal() == 1 || securitySystem->current.getVal() > 0) == 2  // <-- This will always return FALSE, because 0 or 1 never equals 2
      && digitalRead(FrontDoor == LOW)) {
```
...
I don't think this is the behavior you are looking for.

**Reply by adamlilley88** [reply]
I don’t understand why it’s interpreted as >0 ?

I have defined the valid states for the security system as 3(disarmed) 1(away) and 2(night)?

Edit: Yes I understand now, the placement of the close bracket. This has been fixed already, and I am using the AlarmStatus variable to compare against, as in my above post

**Reply by frankonski** [reply]
> I don’t understand why it’s interpreted as >0 ?

false = 0. True = 1. Zero and/or One can't equals 2, hence always False.

Simplifying your code to exclude the PIN logic, what are you trying to achieve?
```
if ((securitySystem->current.getVal() == 1 || securitySystem->current.getVal()) == 2) {
```
or
```
if ((securitySystem->current.getVal() == 1 || securitySystem->current.getVal() == 2)) {
```
? These two lines have quite meaningful differences.

**Comment by adamlilley88** [comment]
Got there in the end!

turns out the error in my code, was down to the fact I was using .getValue() as suggested by the IDE.

Changing to .getVal() fixed it for me. Now have a fully working alarm system on the chip

Thanks for your help pointing me in the right direction and not just doing it for me.
